src/services/jsTyping.ts

```diff
@@ -5,49 +5,190 @@
-                        cachedTypingNames.push(cachedTypingName);
-                        cachedTypingPaths.push(ts.combinePaths(typingsPath, cachedTypingPath));
+            try{
+                let cacheTsdJsonDict = JSON.parse(host.readFile(cacheTsdJsonPath));
+                // The "installed" property in the tsd.json servers as a registry of installed typings. Each item 
+                // of this object has a key of the relative file path, and a value that contains the corresponding
+                // commit hash.
+                if (cacheTsdJsonDict.hasOwnProperty("installed")) {
+                    for (let cachedTypingPath of Object.keys(cacheTsdJsonDict.installed)) {
+                        // Assuming the cachedTypingPath has the format of "[package name]/[file name]"
+                        // Todo: sometimes the package names may not match exactly. For example, in package.json angular
+                        // is written as "angular", however the resolved typing is "angularjs/..", therefore it would never
+                        // match cached version
+                        let cachedTypingName = cachedTypingPath.substr(0, cachedTypingPath.indexOf('/'));
+                        if (inferredTypings.hasOwnProperty(cachedTypingName) && !inferredTypings[cachedTypingName]) {
+                            inferredTypings[cachedTypingName] = ts.combinePaths(typingsPath, cachedTypingPath);
+                        }
                     }
                 }
             }
+            catch(e) { }
         }
         
-        let newTypingNames = typingNames.filter(t => cachedTypingNames.indexOf(t) < 0);
+        let newTypingNames: string[] = [];
+        let cachedTypingPaths: string[] = [];
+        for (let typing in inferredTypings) {
+            if(inferredTypings[typing]) {
+                cachedTypingPaths.push(inferredTypings[typing]);
+            }
+            else {
+                newTypingNames.push(typing);
+            }
+        }
         
-        return { cachedTypingPaths, newTypings: newTypingNames };
+        return { cachedTypingPaths, newTypingNames };
+    }
+    
+    function mergeTypings(typingNames: string[]) {
+        for (let typing of typingNames) {
+            if (!inferredTypings.hasOwnProperty(typing)) {
+                inferredTypings[typing] = undefined;
+            }
+        }
+    }
+
+    function getTypingNamesFromPackageJson(packageJsonPath: string) {
+        if (_host.fileExists(packageJsonPath)) {
+            // Guide against malformed package.json
+            try{
+                let packageJsonDict = JSON.parse(_host.readFile(packageJsonPath));
+                if (packageJsonDict.hasOwnProperty("dependencies")) {
+                    mergeTypings(Object.keys(packageJsonDict.dependencies));
+                }
+            }
+            catch(e) {
+            }
+        }
+    }
+
+    function getTypingNamesFromBowerJson(bowerJsonPath: string) {

```

Comment:
identical to getTypingNamesFromPackageJson?  If so, remove and use getTypingsNamesFromJson()